sa/sa_test.go

```diff
@@ -991,6 +992,117 @@ func TestNewOrderAndAuthzs(t *testing.T) {
+	initialIP, _ := net.ParseIP("17.17.17.17").MarshalText()
+	reg, err := sa.NewRegistration(ctx, &corepb.Registration{
+		Key:       key,
+		InitialIP: initialIP,
+	})
+	test.AssertNotError(t, err, "Couldn't create test registration")
+
+	_, err = sa.NewOrderAndAuthzs(context.Background(), &sapb.NewOrderAndAuthzsRequest{
+		NewAuthzs: []*corepb.Authorization{
+			{
+				Identifier:     "a.com",
+				RegistrationID: reg.Id,
+				Expires:        fc.Now().Add(2 * time.Hour).UnixNano(),
+				Status:         "pending",
+				Challenges:     []*corepb.Challenge{{Token: core.NewToken()}},
+			},
+		},
+	})
+
+	test.AssertErrorIs(t, err, errIncompleteRequest)
+}
+
+func TestNewOrderAndAuthzs_NewAuthzExpectedFields(t *testing.T) {
+	sa, fc, cleanup := initSA(t)
+	defer cleanup()
+
+	// Create a test registration to reference
+	key, _ := jose.JSONWebKey{Key: &rsa.PublicKey{N: big.NewInt(1), E: 1}}.MarshalJSON()
+	initialIP, _ := net.ParseIP("17.17.17.17").MarshalText()
+	reg, err := sa.NewRegistration(ctx, &corepb.Registration{
+		Key:       key,
+		InitialIP: initialIP,
+	})
+	test.AssertNotError(t, err, "Couldn't create test registration")
+
+	expires := fc.Now().Add(time.Hour).UTC().UnixNano()
+	validUntil := fc.Now().Add(time.Hour * -1).UTC().UnixNano()
+	domain := "a.com"
+
+	// Create the order object
+	order := &corepb.Order{
+		RegistrationID: reg.Id,
+		Expires:        expires,
+		Names:          []string{domain},
+	}
+
+	// Create a pending authz that does not yet exist in the database.
+	_, err = sa.NewOrderAndAuthzs(context.Background(), &sapb.NewOrderAndAuthzsRequest{
+		NewAuthzs: []*corepb.Authorization{
+			{
+				Identifier:     domain,
+				RegistrationID: reg.Id,
+				Expires:        expires,
+				Status:         string(core.StatusPending),
+				Challenges:     []*corepb.Challenge{{Token: core.NewToken()}},
+			},
+		},
+		NewOrder: &sapb.NewOrderRequest{
+			RegistrationID: order.RegistrationID,
+			Expires:        order.Expires,
+			Names:          order.Names,
+		},
+	})
+	test.AssertNotError(t, err, "sa.NewOrderAndAuthzs failed")
+
+	// Retrieve the new authz created above and verify that specific fields were
+	// set to their golang zero-value. We store the database output into the &am
+	// authzModel{} holder because we want to check for nil fields. If we were
+	// to use a function such as sa.GetAuthorizations2, then the database output
+	// would be serialized into a protobuf message and the nils would be lost,
+	// rendering potential checking impossible.
+	var am = authzModel{}
+	_ = sa.dbReadOnlyMap.SelectOne(
+		&am,
+		fmt.Sprintf(`SELECT %s FROM authz2 WHERE
+			registrationID = :regID AND
+			expires > :validUntil AND
+			identifierType = :dnsType AND
+			identifierValue = :ident
+			ORDER BY expires ASC
+			LIMIT 1 `, authzFields),
+		map[string]interface{}{
+			"regID":      reg.Id,
+			"validUntil": validUntil,
+			"dnsType":    identifierTypeToUint[string(identifier.DNS)],
+			"ident":      domain,
+		},
```

Comment:
This is a fairly complex SELECT. Instead of doing this, we can select the new authorization directly by its ID if we don't throw away the order object returned by sa.NewOrderAndAuthzs on line 1050. That order object has a `V2Authorizations` field which is a list of database-assigned unique authz2 IDs. So just grab the id from there, and select for it directly, like we do here:
https://github.com/letsencrypt/boulder/blob/9d9a2dddcf0d8c59b0dc85fc4f576aec3a849b43/sa/saro.go#L720

Potentially even simpler: get that ID, and then pass it to `sa.GetAuthorization2`, and inspect the result of that.

Reply:
I tried `sa.GetAuthorization2`, but wanted the raw data including MariaDB null fields out of the database. I have a comment about that attempt at https://github.com/letsencrypt/boulder/pull/6622/files#diff-0b4a90e7036057a14cf4c468d677350e1807b189a1c13ee63bccdb458254bafcR1068-R1073